File: src/main/java/gg/mineral/practice/queue/Queuetype.java
package gg.mineral.practice.queue;

import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

import org.bukkit.inventory.ItemStack;
import org.eclipse.jdt.annotation.Nullable;

import gg.mineral.api.collection.GlueList;
import gg.mineral.api.config.FileConfiguration;
import gg.mineral.practice.arena.Arena;
import gg.mineral.practice.catagory.Catagory;
import gg.mineral.practice.entity.ProfileData;
import gg.mineral.practice.gametype.Gametype;
import gg.mineral.practice.managers.ArenaManager;
import gg.mineral.practice.managers.EloManager;
import gg.mineral.practice.managers.QueuetypeManager;
import gg.mineral.practice.match.data.MatchData;
import gg.mineral.practice.util.SaveableData;
import gg.mineral.practice.util.collection.LeaderboardMap;
import gg.mineral.practice.util.items.ItemStacks;
import gg.mineral.practice.util.messages.CC;
import gg.mineral.server.combat.KnockbackProfile;
import gg.mineral.server.combat.KnockbackProfileList;
import it.unimi.dsi.fastutil.bytes.ByteIterator;
import it.unimi.dsi.fastutil.bytes.ByteOpenHashSet;
import it.unimi.dsi.fastutil.bytes.ByteSet;
import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
import it.unimi.dsi.fastutil.objects.ObjectSet;
import lombok.Getter;

public class Queuetype implements SaveableData {
	final FileConfiguration config = QueuetypeManager.getConfig();

	@Getter
	ItemStack displayItem;
	@Getter
	String name;
	@Getter
	String displayName;
	@Getter
	int slotNumber;
	@Getter
	boolean ranked, community, unranked, botsEnabled;
	final String path;
	@Getter
	KnockbackProfile knockback = null;
	@Getter
	Object2IntOpenHashMap<Gametype> gametypes = new Object2IntOpenHashMap<>();
	@Getter
	ByteSet arenas = new ByteOpenHashSet();
	@Getter
	Object2IntOpenHashMap<Catagory> catagories = new Object2IntOpenHashMap<>();
	@Getter
	private final byte id;
	private long arenaIndex = 0;

	public Queuetype(String name, byte id) {
		this.name = name;
		this.id = id;
		this.path = "Queue." + getName() + ".";
	}

	ByteSet arenaQueue = new ByteOpenHashSet();

	public Gametype randomGametype() {
		List<Gametype> list = new GlueList<>(gametypes.keySet());
		Random rand = new Random();
		return list.get(rand.nextInt(list.size()));
	}

	public Gametype randomGametypeWithBotsEnabled() {
		List<Gametype> list = new GlueList<>(gametypes.keySet()).stream().filter(g -> g.isBotsEnabled())
				.collect(Collectors.toList());
		Random rand = new Random();
		return list.get(rand.nextInt(list.size()));
	}

	// Helper method to filter available arenas based on a Gametype
	public ByteSet filterArenasByGametype(Gametype g) {
		ByteSet filteredArenas = new ByteOpenHashSet(this.arenas);
		filteredArenas
				.retainAll(g.getArenas());
		return filteredArenas;
	}

	public int getGlobalElo(ProfileData profile) {

		ObjectSet<Gametype> set = getGametypes().keySet();

		int sum = 0;

		for (Gametype gametype : set)
			sum += gametype.getElo(profile);

		return Math.round(sum / set.size());
	}

	public List<String> getGlobalLeaderboardLore() {
		List<String> lore = new GlueList<>();

		LeaderboardMap leaderboardMap = EloManager.getGlobalEloLeaderboard(this);

		lore.add(CC.WHITE + "The " + CC.SECONDARY + "global" + CC.WHITE + " elo leaderboard.");
		lore.add(" ");

		for (gg.mineral.practice.util.collection.LeaderboardMap.Entry entry : leaderboardMap.getEntries())
			lore.add(CC.SECONDARY + entry.getKey() + ": " + CC.WHITE + entry.getValue());

		if (lore.size() <= 2)
			lore.add(CC.ACCENT + "No Data");

		return lore;
	}

	@Nullable
	public byte nextArenaId(Gametype g) {
		// Return early if there are no arenas
		if (arenas.isEmpty() || g.getArenas().isEmpty())
			return -1;

		arenaQueue.removeIf(arenaId -> !g.getArenas().contains(arenaId));

		if (arenaQueue.isEmpty())
			arenaQueue.addAll(filterArenasByGametype(g));

		if (arenaQueue.isEmpty())
			return -1;

		ByteIterator iterator = arenaQueue.iterator();

		byte nextArenaId = iterator.nextByte();
		iterator.remove();

		return nextArenaId;
	}

	public byte nextArenaId(MatchData matchData, Gametype g) {

		// If there are no enabled arenas in the MatchData, revert to the other method
		if (matchData.getEnabledArenas().isEmpty())
			return nextArenaId(g);

		// Filter arenas based on Gametype and MatchData
		ByteSet filteredArenas = filterArenasByGametype(g);

		ByteIterator iterator = filteredArenas.iterator();
		while (iterator.hasNext()) {
			byte arenaId = iterator.nextByte();
			if (!matchData.getEnabledArenas().get(arenaId))
				iterator.remove();
		}

		if (filteredArenas.isEmpty())
			return nextArenaId(g);

		// Select a random arena from the filtered list
		int randomIndex = (int) (arenaIndex++ % filteredArenas.size());
		byte selectedArenaId = filteredArenas.toArray(new byte[0])[randomIndex];

		// Remove the selected arena from the main queue to avoid repetition
		arenaQueue.remove(selectedArenaId);

		return selectedArenaId;
	}

	public void setDisplayName(String displayName) {
		this.displayName = displayName;
		save();
	}

	public void setDisplayItem(ItemStack displayItem) {
		this.displayItem = displayItem;
		save();
	}

	public void setSlotNumber(int slotNumber) {
		this.slotNumber = slotNumber;
		save();
	}

	public void setRanked(boolean ranked) {
		this.ranked = ranked;
		save();
	}

	public void setBotsEnabled(boolean bots) {
		this.botsEnabled = bots;
		save();
	}

	public void setCommunity(boolean community) {
		this.community = community;
		save();
	}

	public void setUnranked(boolean unranked) {
		this.unranked = unranked;
		save();
	}

	public void setKnockback(KnockbackProfile knockback) {
		this.knockback = knockback;
		save();
	}

	public void enableArena(Arena arena, boolean enabled) {
		if (enabled) {
			arenas.add(arena.getId());
		} else {
			arenas.remove(arena.getId());
			arenaQueue.remove(arena.getId());
		}
		save();
	}

	public boolean equals(Queuetype q) {
		return q.getName().equalsIgnoreCase(getName());
	}

	public void addCatagory(Catagory catagory, int slot) {
		catagories.put(catagory, slot);
		save();
	}

	public void addGametype(Gametype gametype, int slot) {
		gametypes.put(gametype, slot);
		save();
	}

	@Override
	public void save() {
		config.set(path + "DisplayName", displayName);
		config.set(path + "Elo", ranked);
		config.set(path + "Community", community);
		config.set(path + "Unranked", unranked);
		config.set(path + "Slot", slotNumber);
		config.set(path + "DisplayItem", displayItem);
		config.set(path + "Bots", botsEnabled);

		if (knockback != null)
			config.set(path + "Knockback", knockback.getName());

		for (Arena arena : ArenaManager.getArenas().values())
			config.set(path + "Arenas." + arena.getName(), getArenas().intStream()
					.filter(id -> id == arena.getId()).findFirst().isPresent());

		config.save();
	}

	@Override
	public void load() {
		this.displayItem = config.getItemstack(path + "DisplayItem",
				ItemStacks.DEFAULT_QUEUETYPE_DISPLAY_ITEM);
		this.displayName = config.getString(path + "DisplayName", getName());
		this.slotNumber = config.getInt(path + "Slot", 8);
		this.ranked = config.getBoolean(path + "Elo", false);
		this.community = config.getBoolean(path + "Community", false);
		this.unranked = config.getBoolean(path + "Unranked", false);
		this.botsEnabled = config.getBoolean(path + "Bots", false);
		String kbprofile = config.getString(path + "Knockback", "");

		if (!kbprofile.isEmpty()) {
			this.knockback = KnockbackProfileList.getKnockbackProfileByName(kbprofile);
		}

		for (Arena a : ArenaManager.getArenas().values())
			if (config.getBoolean(path + "Arenas." + a.getName(), false))
				arenas.add(a.getId());

	}

	@Override
	public void setDefaults() {
		this.displayItem = ItemStacks.DEFAULT_QUEUETYPE_DISPLAY_ITEM;
		this.displayName = getName();
		this.slotNumber = 8;
		this.ranked = false;
		this.botsEnabled = false;
	}

	@Override
	public void delete() {
		config.remove("Queue." + getName());
		config.save();
	}
}

File: src/main/kotlin/gg/mineral/practice/queue/Queuetype.kt
package gg.mineral.practice.queue

import gg.mineral.api.collection.GlueList
import gg.mineral.practice.arena.Arena
import gg.mineral.practice.category.Category
import gg.mineral.practice.entity.ProfileData
import gg.mineral.practice.gametype.Gametype
import gg.mineral.practice.managers.*
import gg.mineral.practice.managers.ArenaManager.getArenaByName
import gg.mineral.practice.util.config.BoolProp
import gg.mineral.practice.util.config.IntProp
import gg.mineral.practice.util.config.ItemStackProp
import gg.mineral.practice.util.config.StringProp
import gg.mineral.practice.util.config.yaml.FileConfiguration
import gg.mineral.practice.util.items.ItemStacks
import gg.mineral.practice.util.messages.CC
import gg.mineral.server.combat.KnockbackProfile
import gg.mineral.server.combat.KnockbackProfileList
import it.unimi.dsi.fastutil.bytes.ByteOpenHashSet
import it.unimi.dsi.fastutil.bytes.ByteSet
import it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap
import it.unimi.dsi.fastutil.objects.Object2IntMap
import it.unimi.dsi.fastutil.objects.ObjectSortedSet
import java.util.*
import java.util.concurrent.CompletableFuture

class Queuetype(val name: String, val id: Byte) {
    val config: FileConfiguration = QueuetypeManager.config
    val path: String = "Queue.$name."
    var displayItem by ItemStackProp(config, "$path.DisplayItem", ItemStacks.DEFAULT_QUEUETYPE_DISPLAY_ITEM)
    var displayName by StringProp(config, "$path.DisplayName", name)
    var slotNumber by IntProp(config, "$path.Slot", 8)
    var ranked by BoolProp(config, "$path.Elo", false)
    var community by BoolProp(config, "$path.Community", false)
    var unranked by BoolProp(config, "$path.Unranked", false)
    var botsEnabled by BoolProp(config, "$path.Bots", false)
    private var knockbackName by StringProp(config, "$path.Knockback", "")
    var knockback: KnockbackProfile?
        get() = KnockbackProfileList.getKnockbackProfileByName(knockbackName)
        set(value) {
            knockbackName = value?.name ?: ""
            config["$path.Knockback"] = knockbackName
        }


    val menuEntries: Object2IntLinkedOpenHashMap<QueuetypeMenuEntry> by lazy {
        val map = object : Object2IntLinkedOpenHashMap<QueuetypeMenuEntry>() {
            override fun put(key: QueuetypeMenuEntry, value: Int): Int {
                val slot = super.put(key, value)
                sortByValues()
                return slot
            }

            private fun sortByValues() {
                val sorted: Object2IntMap<QueuetypeMenuEntry> = object2IntEntrySet().stream()
                    .sorted(Comparator.comparingInt { obj: Object2IntMap.Entry<QueuetypeMenuEntry?> -> obj.intValue })
                    .collect(
                        { Object2IntLinkedOpenHashMap() },
                        { map: Object2IntLinkedOpenHashMap<QueuetypeMenuEntry>, entry: Object2IntMap.Entry<QueuetypeMenuEntry> ->
                            map.put(
                                entry.key,
                                entry.intValue
                            )
                        },
                        { obj: Object2IntLinkedOpenHashMap<QueuetypeMenuEntry>, m: Object2IntLinkedOpenHashMap<QueuetypeMenuEntry> ->
                            obj.putAll(m)
                        })

                clear()

                for (entry in sorted.object2IntEntrySet()) super.put(entry.key, entry.intValue)
            }

            @Deprecated(
                "Deprecated in Java",
                ReplaceWith("super.getInt(key)", "it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap")
            )
            override fun get(key: QueuetypeMenuEntry) = super.getInt(key)

            @Deprecated(
                "Deprecated in Java",
                ReplaceWith("super.removeInt(key)", "it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap")
            )
            override fun remove(key: QueuetypeMenuEntry) = super.removeInt(key)
        }

        // TODO: move to queuetype.yml
        val gametypeConfig = GametypeManager.config

        for (gametype in GametypeManager.gametypes.values) {
            if (!gametypeConfig.getBoolean("Gametype.${gametype.name}.$name.Enabled", false)) continue
            map[gametype] = gametypeConfig.getInt("Gametype.${gametype.name}.$name.Slot", 0)
        }

        // TODO: move to queuetype.yml
        val categoryConfig = CategoryManager.config

        for (category in CategoryManager.categories.values) {
            if (!categoryConfig.getBoolean("Category.${category.name}.$name.Enabled", false)) continue
            map[category] = categoryConfig.getInt("Category.${category.name}.$name.Slot", 0)
        }

        map
    }


    val arenas by lazy {
        val set = ByteOpenHashSet()
        for (a in ArenaManager.arenas.values) if (config.getBoolean(path + "Arenas." + a.name, false)) set.add(a.id)
        set
    }

    fun randomGametype(): Gametype {
        val list = GlueList<Gametype>()

        for (menuEntry in menuEntries.keys) if (menuEntry is Gametype) list.add(menuEntry)

        val rand = Random()
        return list[rand.nextInt(list.size)]
    }

    fun randomGametypeWithBotsEnabled(): Gametype {
        val list = GlueList<Gametype>()

        for (menuEntry in menuEntries.keys) if (menuEntry is Gametype && menuEntry.botsEnabled) list.add(menuEntry)

        val rand = Random()
        return list[rand.nextInt(list.size)]
    }

    // Helper method to filter available arenas based on a Gametype
    fun filterArenasByGametype(g: Gametype): ByteSet {
        val filteredArenas = ByteOpenHashSet(this.arenas)
        filteredArenas
            .retainAll(g.arenas)
        return filteredArenas
    }

    fun getGlobalElo(profile: ProfileData): CompletableFuture<Int> {
        val set: ObjectSortedSet<QueuetypeMenuEntry> = menuEntries.keys

        val futures = mutableListOf<CompletableFuture<Int>>()

        for (menuEntry in set) if (menuEntry is Gametype) {
            futures.add(menuEntry.getElo(profile))
        }

        if (futures.isEmpty()) return CompletableFuture.completedFuture(1000)

        return CompletableFuture.allOf(*futures.toTypedArray()).thenApply {
            var sum = 0
            var divisor = 0

            for (future in futures) {
                sum += future.join()
                divisor++
            }
            if (divisor == 0) return@thenApply 1000
            sum / divisor
        }
    }

    val globalLeaderboardLore: List<String>
        get() {
            val lore = GlueList<String>()

            val leaderboardMap = EloManager.getGlobalEloLeaderboard(this)

            lore.add(CC.WHITE + "The " + CC.SECONDARY + "global" + CC.WHITE + " elo leaderboard.")
            lore.add(" ")

            for (entry in leaderboardMap.entries) lore.add(CC.SECONDARY + entry.key + ": " + CC.WHITE + entry.value)

            if (lore.size <= 2) lore.add(CC.ACCENT + "No Data")

            return lore
        }

    fun enableArena(arena: Arena, enabled: Boolean) {
        if (enabled) arenas.add(arena.id)
        else arenas.remove(arena.id)
        config[path + "Arenas." + arena.name] = arenas.contains(arena.id) && getArenaByName(arena.name) != null
    }

    override fun equals(other: Any?): Boolean {
        if (other !is Queuetype) return false
        return other.name.equals(name, ignoreCase = true)
    }

    fun addMenuEntry(menuEntry: QueuetypeMenuEntry, slot: Int) {
        menuEntries.put(menuEntry, slot)

        //TODO: move to queuetype.yml
        val gametypeConfig = GametypeManager.config

        if (menuEntry is Gametype) {
            gametypeConfig["Gametype.${menuEntry.name}.$name.Enabled"] = true
            gametypeConfig["Gametype.${menuEntry.name}.$name.Slot"] = slot
        }

        //TODO: move to queuetype.yml
        val categoryConfig = QueuetypeManager.config

        if (menuEntry is Category) {
            categoryConfig["Category.${menuEntry.name}.$name.Enabled"] = true
            categoryConfig["Category.${menuEntry.name}.$name.Slot"] = slot
        }
    }

    fun removeMenuEntry(menuEntry: QueuetypeMenuEntry) {
        menuEntries.removeInt(menuEntry)
        //TODO: move to queuetype.yml
        val gametypeConfig = GametypeManager.config

        if (menuEntry is Gametype)
            gametypeConfig["Gametype.${menuEntry.name}.$name.Enabled"] = false

        //TODO: move to queuetype.yml
        val categoryConfig = QueuetypeManager.config

        if (menuEntry is Category)
            categoryConfig["Category.${menuEntry.name}.$name.Enabled"] = false
    }

    fun delete() {
        config.remove("Queue.$name")
        config.save()
    }

    override fun hashCode() = javaClass.hashCode()
}
